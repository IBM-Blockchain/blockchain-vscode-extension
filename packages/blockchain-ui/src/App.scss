@import '@carbon/colors/scss/colors';

#root{
    .app-container__tutorial-page {
        background-color: $gray-10;
        position: absolute;
        width: 100%;
        z-index: -1;
        box-shadow: inset 0px 164x #FFFFFF;
        top: 0;
        left: 0;

    }
}
