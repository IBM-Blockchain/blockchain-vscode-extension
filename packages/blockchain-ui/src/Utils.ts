import IVSCode from './interfaces/IVSCode';
declare const vscode: IVSCode;

class Utils {

    static changeRoute(newPath: string): void {
        dispatchEvent(new MessageEvent('message', {
            data: {
                path: newPath
            }
        }));
    }

    static postToVSCode(message: { command: string, data?: any }): void {

        if (!(window as any)['Cypress']) {
            vscode.postMessage(message);
        } else {
            window.postMessage(message, '*');
        }
    }

}

export default Utils;
