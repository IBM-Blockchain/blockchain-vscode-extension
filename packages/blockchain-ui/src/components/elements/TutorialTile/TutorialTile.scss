@import '@carbon/colors/scss/colors';
@import '@carbon/layout/scss/_spacing';

.tab-container {
    display: flex;
    flex-direction: column;
    max-width: 70%;
    background-color: white;
    margin-bottom: $spacing-03;
}

.tutorial-title {
    font-family: 'IBMPlexSans-SemiBold', sans-serif;
    font-size: 14px;
    letter-spacing: 0.16px;
    line-height: 18px;
    color: $gray-100;
    margin: $spacing-05;
}

.time-container {
    display: inline-flex;
    font-family: 'IBMPlexSans', sans-serif;
    color: $gray-80;
    letter-spacing: 0.32px;
    line-height: $spacing-05;
    margin-left: $spacing-05;
}

.icon {
    margin-right: $spacing-03;
    fill: black;
}

.time-text {
    align-items: center;
    font-size: 12px;
}

.tutorial-objectives {
    min-width: 60%;
    padding-left: 40px;
    padding-top: $spacing-05;
}

.objective {
    color: $gray-70;
    font-family: 'IBMPlexSans', sans-serif;
    font-size: 12px;
    letter-spacing: 0.32px;
    text-align: left;
}

.button {
    max-width: 170px;
}

.pdf-button {
    padding-right: 16px;
    border: none;
    color: $blue-60;
}

.button-container {
    margin-left: auto;
}
