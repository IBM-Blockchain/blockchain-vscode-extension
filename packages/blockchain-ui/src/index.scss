@import '@carbon/colors/scss/colors';
@import '@carbon/themes/scss/themes';
@import '@carbon/type/scss/type';
@import '@carbon/type/scss/font-face/_mono';
@import '@carbon/type/scss/font-face/_sans';
@import './mixins/theme-mixins.scss';

@include carbon--type-reset();
@include carbon--font-face-mono();
@include carbon--font-face-sans();

html {

    .vscode-dark {
        @include carbon--theme($carbon--theme--g90) {

            #root {
                // Color will default to #000000 otherwise
                color: $text-01;        
                @include select-dark;
            }

            // text inputs on the create transaction screen
            #create-txn-form {
                @include text-inputs-dark;
            }

            // data tables on home page
            #recent-txns-table, #saved-txns-table {
                @include data-table-dark;
            }

            // InlineSelect components
            #smart-contract-select-wrapper {
                @include inline-select-dark;
            }

            // Checkbox
            @include checkbox-dark;

            //panel containing transaction output
            #output-panel {
                @include output-panel-dark;
            }
        }
    }

    .vscode-light {
        @include carbon--theme($carbon--theme--g10) {
            
            #root {
                color: $text-01;
            }
        }
    }
}
