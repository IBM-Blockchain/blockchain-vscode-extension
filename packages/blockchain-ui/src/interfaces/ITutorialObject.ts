interface ITutorialObject {
    title: string;
    series: string;
    length: string;
    firstInSeries?: boolean;
    file: string;
    objectives: string[];
    badge?: boolean;
}

export default ITutorialObject;
