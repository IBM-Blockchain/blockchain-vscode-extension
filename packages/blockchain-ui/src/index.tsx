import * as React from 'react';
import * as ReactDOM from 'react-dom';
import App from './App';
import './index.scss';

const app: any = ReactDOM.render(
    <App/>,
    document.getElementById('root')
);

// @ts-ignore
if (window.Cypress) {
    // @ts-ignore
    window.app = app;
}
