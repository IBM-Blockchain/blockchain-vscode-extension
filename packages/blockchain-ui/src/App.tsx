import * as React from 'react';
import './App.scss';
import { HashRouter as Router, Route, Redirect } from 'react-router-dom';
import TransactionHome from './components/TransactionHome/TransactionHome';
import TransactionCreate from './components/TransactionCreate/TransactionCreate';
import ISmartContract from './interfaces/ISmartContract';
import Utils from './Utils';

interface AppState {
    redirectPath: string;
    gatewayName: string;
    activeSmartContract: ISmartContract;
    smartContracts: Array<ISmartContract>;
    transactionOutput: string;
}

class App extends React.Component<{}, AppState> {
    constructor(props: {}) {
        super(props);
        this.state = {
            redirectPath: '',
            gatewayName: '',
            activeSmartContract: {
                name: '',
                version: '',
                channel: '',
                label: '',
                transactions: [],
                namespace: ''
            },
            smartContracts : [],
            transactionOutput: ''
        };
        this.switchSmartContract = this.switchSmartContract.bind(this);
        this.postMessageHandler = this.postMessageHandler.bind(this);
    }

    componentDidMount(): void {
        window.addEventListener('message', (event: MessageEvent) => {
            if (event.data.output) {
                this.setState({
                    transactionOutput: event.data.output
                });
            } else {
                this.setState({
                    redirectPath: event.data.path,
                    gatewayName: event.data.state ? event.data.state.gatewayName : this.state.gatewayName,
                    activeSmartContract: event.data.state ? event.data.state.activeSmartContract : this.state.activeSmartContract,
                    smartContracts: event.data.state ? event.data.state.smartContracts : this.state.smartContracts,
                });
            }
        });
    }

    switchSmartContract(newActiveContractLabel: string): void {
        const smartContracts: Array<ISmartContract> = this.state.smartContracts;
        this.setState({
            activeSmartContract: smartContracts.find((obj: ISmartContract) => obj.label === newActiveContractLabel) as ISmartContract,
            smartContracts: smartContracts
        });
    }

    postMessageHandler(command: string, data?: any): void {
        if (data === undefined) {
            data = {
                gatewayName: this.state.gatewayName,
                activeSmartContract: this.state.activeSmartContract,
                smartContracts: this.state.smartContracts
            };
        }

        Utils.postToVSCode({
            command: command,
            data: data
        });
    }

    render(): JSX.Element {
        if (this.state.redirectPath === '') {
            return <div></div>;
        } else {
            return (
                <Router>
                    <div>
                        <Route render={(): JSX.Element => <Redirect push to={this.state.redirectPath}/>}></Route>
                        <Route exact path='/transaction' render={(): JSX.Element =>
                            <TransactionHome gatewayName={this.state.gatewayName} activeSmartContract={this.state.activeSmartContract} smartContracts={this.state.smartContracts}
                                switchSmartContract={this.switchSmartContract} postMessageHandler={this.postMessageHandler}/>}>
                        </Route>
                        <Route exact path='/transaction/create' render={(): JSX.Element =>
                            <TransactionCreate activeSmartContract={this.state.activeSmartContract} transactionOutput={this.state.transactionOutput}
                                postMessageHandler={this.postMessageHandler}/>}>
                        </Route>
                    </div>
                </Router>
            );
        }
    }
}

export default App;
