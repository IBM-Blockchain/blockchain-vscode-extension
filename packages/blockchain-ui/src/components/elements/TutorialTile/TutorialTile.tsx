import React, { Component } from 'react';
import './TutorialTile.scss';
import { Timer16 } from '@carbon/icons-react';
import { Button } from 'carbon-components-react';

interface IProps {
    tutorialObject: any;
}

class TutorialTile extends Component <IProps> {

    render(): JSX.Element {
        return (
            <div className='tab-container'>
                <div className='tutorial-title'>
                    {this.props.tutorialObject.title}
                </div>
                <div className='time-container'>
                    <Timer16 className='icon'></Timer16>
                    <span className='time-text'>{this.props.tutorialObject.length}</span>
                </div>
                <div className='tutorial-objectives'>
                    <p className='objective'>Learn about what blockchain is and why its important</p>
                    <p className='objective'>Learn about the Linux Foundation Hyperledger Project and Hyperledger Fabric</p>
                    <p className='objective'>Learn about IBM Blockchain Platform and the VS Code extension</p>
                </div>
                <div className='button-container'>
                    <Button className='pdf-button' kind='ghost' size='default'>Download as PDF</Button>
                    {this.props.tutorialObject.firstInSeries === true ?
                        <Button className='button' kind='primary' size='default'>Open tutorial</Button> :
                        <Button className='button' kind='secondary' size='default'>Open tutorial</Button>
                    }
                </div>
            </div>
        );
    }
}

export default TutorialTile;
