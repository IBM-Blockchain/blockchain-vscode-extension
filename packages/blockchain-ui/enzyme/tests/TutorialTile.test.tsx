import React from 'react';
import renderer from 'react-test-renderer';
import { mount, ReactWrapper } from 'enzyme';
import chai from 'chai';
import sinon from 'sinon';
import sinonChai from 'sinon-chai';

import TutorialTile from '../../src/components/elements/TutorialTile/TutorialTile';
import ITutorialObject from '../../src/interfaces/ITutorialObject';
import Utils from '../../src/Utils';
import { ExtensionCommands } from '../../src/ExtensionCommands';

chai.should();
chai.use(sinonChai);

// tslint:disable no-unused-expression
interface IProps {
    tutorialObject: ITutorialObject;
}

describe('TutorialTile component', () => {
    let mySandBox: sinon.SinonSandbox;
    const tutorialObject: ITutorialObject = {
            title: 'a1',
            series: 'my series',
            firstInSeries: true,
            length: '4 weeks',
            objectives: [
                'objective 1',
                'objective 2',
                'objective 3'
            ],
            file: 'some/file/path'
    };

    beforeEach(() => {
        mySandBox = sinon.createSandbox();
    });

    it('should render the expected snapshot', () => {
        const component: any = renderer
            .create(<TutorialTile tutorialObject={tutorialObject}/>)
            .toJSON();
        expect(component).toMatchSnapshot();
    });

    it('should render a primary button if tutorial is first in series', () => {
        const component: ReactWrapper<IProps> = mount(<TutorialTile tutorialObject={tutorialObject}/>);
        component.find('button').at(1).hasClass('bx--btn--primary').should.equal(true);
    });

    it('should render a secondary button if tutorial is not first in series', () => {
        const anotherTutorialObject: ITutorialObject = {
            title: 'a4',
            series: 'my series',
            length: '3 weeks',
            objectives: [
                'objective 1',
                'objective 2',
                'objective 3'
            ],
            file: 'some/file/path'
        };
        const component: ReactWrapper<IProps> = mount(<TutorialTile tutorialObject={anotherTutorialObject}/>);
        component.find('button').at(1).hasClass('bx--btn--secondary').should.equal(true);
    });

    it(`should post a message to VS Code when the 'Open tutorial' button is clicked`, () => {
        const postToVSCodeStub: sinon.SinonStub = mySandBox.stub(Utils, 'postToVSCode').resolves();
        const component: ReactWrapper<IProps> = mount(<TutorialTile tutorialObject={tutorialObject}/>);
        component.find('button').at(1).simulate('click');
        postToVSCodeStub.should.have.been.calledOnceWithExactly({
            command: ExtensionCommands.OPEN_REACT_TUTORIAL_PAGE,
            data: [
                tutorialObject.series,
                tutorialObject.title
            ]
        });
    });

    it('should show badge available if the tutorial has a badge available', () => {
        const anotherTutorialObject: any = {
            title: 'a4',
            length: '3 weeks',
            badge: true,
            file: 'some/file/path',
            objectives: [
                'objective 1',
                'objective 2',
                'objective 3'
            ]
        };
        const component: ReactWrapper<IProps> = mount(<TutorialTile tutorialObject={anotherTutorialObject}/>);
        component.text().includes('Badge available').should.be.true;
    });

});
