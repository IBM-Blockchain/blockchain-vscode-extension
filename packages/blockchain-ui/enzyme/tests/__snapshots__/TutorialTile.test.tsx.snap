// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TutorialTile component should render the expected snapshot 1`] = `
<div
  className="tab-container"
>
  <div
    className="tutorial-title"
  >
    a1
  </div>
  <div
    className="time-container"
  >
    <svg
      aria-hidden={true}
      className="icon"
      focusable="false"
      height={16}
      preserveAspectRatio="xMidYMid meet"
      viewBox="0 0 32 32"
      width={16}
      xmlns="http://www.w3.org/2000/svg"
    >
      <rect
        height="9"
        width="2"
        x="15"
        y="11"
      />
      <rect
        height="2"
        width="6"
        x="13"
        y="2"
      />
      <path
        d="M28,9,26.58,7.59,24.33,9.84a10.94,10.94,0,1,0,1.18,1.65ZM16,26a9,9,0,1,1,9-9A9,9,0,0,1,16,26Z"
      />
    </svg>
    <span
      className="text"
    >
      4 weeks
    </span>
  </div>
  <div
    className="tutorial-objectives"
  >
    <p
      className="objective"
    >
      objective 1
    </p>
    <p
      className="objective"
    >
      objective 2
    </p>
    <p
      className="objective"
    >
      objective 3
    </p>
  </div>
  <div
    className="button-container"
  >
    <button
      className="pdf-button bx--btn bx--btn--ghost"
      disabled={false}
      tabIndex={0}
      type="button"
    >
      Download as PDF
    </button>
    <button
      className="button bx--btn bx--btn--primary"
      disabled={false}
      onClick={[Function]}
      tabIndex={0}
      type="button"
    >
      Open tutorial
    </button>
  </div>
</div>
`;
